pythran/optimizations/expand_boolean_operator.py

```diff
@@ -0,0 +1,174 @@
+"""
+expand boolean logical operators
+"""
+
+from typing import Union
+import gast as ast
+from pythran.passmanager import Transformation
+
+def _is_bool_call(node: ast.Call) -> bool:
+    if not isinstance(node, ast.Call):
+        return False
+    func = node.func
+    return len(node.args) == 1 and not node.keywords and\
+        isinstance(func, ast.Attribute) and \
+        func.attr == "bool" and \
+        isinstance(func.value, ast.Name) and \
+        func.value.id == "builtins"
+
+def _boolean_BoolOp(op: ast.BoolOp):
+    for i, val in enumerate(op.values):
+        if _is_bool_call(val):
+            continue
+        op.values[i] = _boolean_stmt(val)
+    return op
+
+def _boolean_stmt(stmt: ast.AST) -> ast.Call:
+    if isinstance(stmt, ast.BoolOp):
+        return _boolean_BoolOp(stmt)
+    return ast.Call(
+        func=ast.Attribute(
+            value=ast.Name(
+                id='builtins', ctx=ast.Load(),
+                annotation=None, type_comment=None
+            ),
+            attr='bool',
+            ctx=ast.Load()
+        ),
+        args=[stmt],
+        keywords=[]
+    )
+
+
+class ExpandBooleanOperator(Transformation):
+
+    """Expand bool() to each logical element of BoolOp node.
+
+    This optimization should work after ExpandBuiltins
+    
+    >>> import gast as ast
+    >>> from pythran import passmanager, backend
+    >>> code = '''
+    ... def foo(a, b, c, d):
+    ...     if builtins.bool(a and b) or builtins.bool(c or d):
+    ...         return b or d
+    ...     elif builtins.bool( bar(c or a) or bar(b and d) ):
+    ...         return a
+    ...     else:
+    ...         return b and d'''
+    >>> node = ast.parse(code)
+    >>> pm = passmanager.PassManager("test")
+    >>> _, node = pm.apply(ExpandBooleanOperator, node)
+    >>> print(pm.dump(backend.Python, node))
+    def foo(a, b, c, d):
+        if ((builtins.bool(a) and builtins.bool(b)) or (builtins.bool(c) or builtins.bool(d))):
+            return (b or d)
+        elif (builtins.bool(bar((c or a))) or builtins.bool(bar((b and d)))):
+            return a
+        else:
+            return (b and d)
+    """
+
+    def __init__(self):
+        super(ExpandBooleanOperator, self).__init__()
+        self._bool_call = False
```

Comment:
Use stateful transformation like you do makes the code harder to follow. I think you could (non recursively) detect if a call is a call to builtin.bool, then swicth on the argument: if it's a `BoolOp` node  just distribute `builtins.bool` over the arguments.

But maybe I'm missing something?